### DISCUSSION TITLE
geom_brick_stack as alternative to geom_dotplot

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#46] [General]
Here's a starting point for a new brick stack Stat, that might have some better properties, but similar usage to geom_dotplot. Motivated by into stats class, question #39. I think it would be good to have a weight variable allowed for mapping, and I'd like to explore using GeomBar as a jumping off point instead in the future.  

``` r
library(tidyverse)
compute_group_bricks <- function(data, scales){
  
  data %>% 
    mutate(id = row_number()) %>% 
    mutate(y = id - .5)  
  
}

StatBricks <- ggproto("StatBricks", Stat,
                      compute_group = compute_group_bricks)

palmerpenguins::penguins |>
  filter(sex == "female") |>
  ggplot() + 
  aes(x = species) + 
  geom_tile(stat = StatBricks, 
            width = .2, 
            color = "grey35", 
            fill = "white")
```

![](https://i.imgur.com/oX9tw7y.png)

``` r

palmerpenguins::penguins |>
  filter(sex == "female") |>
  ggplot() + 
  aes(x = species) + 
  geom_dotplot()
#> Bin width defaults to 1/30 of the range of the data. Pick better value with
#> `binwidth`.
```

![](https://i.imgur.com/wk5QL6k.png)

<sup>Created on 2024-10-08 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by yjunechoe** [comment]
A "brick plot" looks like an interesting alternative to dotplot 👀 

Might also be worth noting that `StatBricks` + `GeomTile` + `PositionIdentity` seems equivalent to `StatIdentity` + `GeomCol` + `PositionStack`:

```r
library(tidyverse)
compute_group_bricks <- function(data, scales){
  
  data %>% 
    mutate(id = row_number()) %>% 
    mutate(y = id - .5)  
  
}

StatBricks <- ggproto("StatBricks", Stat,
                      compute_group = compute_group_bricks)

palmerpenguins::penguins |>
  filter(sex == "female") |>
  ggplot() + 
  aes(x = species) + 
  geom_tile(stat = StatBricks, 
            width = .2, 
            color = "grey35", 
            fill = "white")
```

![image](https://github.com/user-attachments/assets/78deefb5-dde4-48cd-8273-a10a3ed3ac09)

```r
palmerpenguins::penguins |>
  filter(sex == "female") |>
  ggplot() + 
  aes(x = species, y = .5) + 
  geom_col(width = .2, 
           linewidth = .2,
           color = "grey35", 
           fill = "white")
```

![image](https://github.com/user-attachments/assets/02c67892-25c5-4a2f-8efc-a4bb0b3e3eb4)

I think this is a place where "on principle" I agree with ggplot's design that this behavior should be handled by Position, not by Stat, but the StatBrick compute method is indeed a simple example to digest!

**Reply by EvaMaeRey** [reply]
hmm!!  I would not have thought to do that with geom_col, but it makes sense (in a fleeting way - I keep having to remind myself of the position stack default)!

**Comment by teunbrand** [comment]
This will probably ruin the simplicity, but dotplots form these histogram like shapes when used with continuous x. If you could get this with tiles as well that'd be pretty cool. However, `geom_dotplot()` has always been a bit of an awkward duck in the pond.

``` r
library(tidyverse)
compute_group_bricks <- function(data, scales){
  
  data %>% 
    mutate(id = row_number()) %>% 
    mutate(y = id - .5)  
  
}

StatBricks <- ggproto("StatBricks", Stat,
                      compute_group = compute_group_bricks)

p <- palmerpenguins::penguins |>
  filter(sex == "female") |>
  ggplot() + 
  aes(x = bill_length_mm) 

p + geom_dotplot(dotsize = 0.5)
#> Bin width defaults to 1/30 of the range of the data. Pick better value with
#> `binwidth`.
```

![](https://i.imgur.com/t2uuopS.png)

``` r

p + geom_tile(stat = StatBricks, 
              width = .2, 
              color = "grey35", 
              fill = "white")
```

![](https://i.imgur.com/S5fW2Tv.png)

<sup>Created on 2024-10-08 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
I like brick histograms!  

What you have there...

p + geom_tile(stat = StatBricks, 
              width = .2, 
              color = "grey35", 
              fill = "white")

...is interesting to think about, but probably no plot anyone wants to make.  

A path w/o extension is to group by row number w/ geom_histogram...

``` r
library(tidyverse)

palmerpenguins::penguins |>
  mutate(id = row_number()) |>
  ggplot() + 
  aes(x = bill_depth_mm) + 
  geom_histogram(color = "white",
                 linewidth = .2) + 
  aes(group = id) + 
  aes(fill = sex)
#> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
#> Warning: Removed 2 rows containing non-finite outside the scale range
#> (`stat_bin()`).
```

![](https://i.imgur.com/UABQ8qd.png)

<sup>Created on 2024-10-08 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>


This is close to how I get to the GeomBar/StatCount jumping off point idea.  It feels like you should be able to just add a little to the existing compute to get grouping after the row index.

**Reply by yjunechoe** [reply]
That's interesting to think about and FWIW, you could get the intended visual if you arrange the data, so that the tiles stack in the correct order:

```r
library(tidyverse)

palmerpenguins::penguins |>
  arrange(sex) |>
  mutate(id = row_number()) |>
  ggplot() + 
  aes(x = bill_depth_mm) + 
  geom_histogram(color = "white",
                 linewidth = .2) + 
  aes(group = id) + 
  aes(fill = sex)
```

![image](https://github.com/user-attachments/assets/a6eecf87-693c-44eb-afb4-db254eb86014)

I know this is a mock-up, but I see that the problem with doing this using `geom_histogram()`/`StatBin` is that you're calling compute for every single tile, just to not do anything meaningful with it:

```r
ggtrace::inspect_n(last_plot(), StatBin$compute_group)
#> [1] 342
```

**Reply by yjunechoe** [reply]
Another (albeit uglier out of the box) alternative is to hand off the relevant computation to the scale (`stale_x_binned()`) instead:

```r
library(tidyverse)

palmerpenguins::penguins |>
  ggplot(aes(bill_depth_mm, fill = sex)) +
  geom_col(aes(y = 1), color = "white") +
  scale_x_binned(
    n.breaks = 30,
    guide = guide_axis(check.overlap = TRUE)
  )
#> Warning: Removed 2 rows containing missing values or values outside the scale range
#> (`geom_col()`).
```

![image](https://github.com/user-attachments/assets/92bd6606-e112-49e1-9293-7135bb2b6853)

**Reply by EvaMaeRey** [reply]
This geom_col X scale_x_binned is interesting (I'm actually surpised by the behavior)!  Your sorted colors are prettier, of course, but in general I don't like stacked histograms because my default interpretation is that they are not stacked but have a common baseline, so I almost prefer the unsorted version because it reminds you that the lower histogram isn't sitting on top of data..

**Reply by EvaMaeRey** [reply]
I need to look at @mjskay's ggdist dotplots!

**Reply by mjskay** [reply]
On this example, ggdist::geom_dots could be something like:

```r
library(tidyverse)
library(ggdist)

palmerpenguins::penguins |>
  ggplot(aes(x = bill_depth_mm, fill = sex)) +
  geom_dots(group = NA)
```
<img width="1575" height="973" alt="image" src="https://github.com/user-attachments/assets/a0e2f503-66c5-4dea-b986-c6e755eaf36c" />

Or:

```r
palmerpenguins::penguins |>
  ggplot(aes(x = bill_depth_mm, fill = sex, order = sex)) +
  geom_dots(group = NA)
```
<img width="1575" height="973" alt="image" src="https://github.com/user-attachments/assets/369135c9-5e5a-4293-aea3-654d0b637759" />

(you can also use the `subguide` option if you want to add labels of counts).

At one point I starting working on weighted dotplots for ggdist but then realized they are basically just stacked bars, as evidenced by this thread :).

**Comment by EvaMaeRey** [comment]
Sketching out basic brickstack layer and friends for looking at proportions as the balancing point for stacks! for #39.  Brick stack does not adapt to continuous variables, but #39 has pretty limited goals.  Maybe separately can tackle layer_brickwall() possibly just following @yjunechoe geom_col X scale_x_binned...   https://github.com/EvaMaeRey/ggstats101 w/ statexpress shorthands.

``` r
donor |>
  ggplot() +
  aes(x = Choice |> fct_rev()) +
  # 1. geom_brickstack() -- show counts in a count-y way (i.e. bricks!), 
  # with a good amount of space in-between stacks
  qlayer(stat = qstat(compute_group_bricks,
                      default_aes = aes(width = after_stat(.2))),
         geom = qproto_update(GeomTile, 
                              aes(color = "whitesmoke")) ) +
  # 2. geom_brickstack_label() -- label stacks
  qlayer(stat = qproto_update(StatCount, aes(label = after_stat(count),
                                             vjust = after_stat(0))),
             geom = GeomText) +
  # 3. geom_xrange(), show scale, range at y is zero
  qlayer(stat = qstat_panel(compute_scale),
             geom = GeomSegment) +
  # 4. geom_prop(), show prop, i.e. balancing point
  qlayer(stat = qstat_panel(compute_xmean_at_y0,
                            default_aes = aes(vjust = after_stat(1),
                                              label = after_stat("^"))),
         geom = qproto_update(GeomText, aes(size = 6))) + 
  # 5. geom_prop_label(), labeling prop, balancing point
  qlayer(stat = qstat_panel(compute_xmean_at_y0,
                            default_aes = aes(label = after_stat(round(x - 1, 2)),
                                              vjust = after_stat(0))),
         geom = qproto_update(GeomLabel, aes(fill = NA, label.size = NA))) ->
proportion_balance_plot  
  
proportion_balance_plot
```


![image](https://github.com/user-attachments/assets/441b20aa-2a75-4e9a-b32c-9aa8b9d667b0)
